tensorboard/data/experimental/experiment_from_dev_test.py

```diff
@@ -0,0 +1,250 @@
+        train_accuracies = list(dataframe.loc["train", ("value", "accuracy")])
+        self.assertAllClose(train_accuracies, 1.0 / (10.0 - np.arange(0, 10)))
+        test_losses = list(dataframe.loc["test", ("value", "loss")])
+        self.assertAllClose(test_losses, -1.0 / (np.arange(0, 10) + 1.0))
+        test_accuracies = list(dataframe.loc["test", ("value", "accuracy")])
+        self.assertAllClose(test_accuracies, -1.0 / (10.0 - np.arange(0, 10)))
+
+        # Check wall_times.
+        train_loss_wall_times = list(
+            dataframe.loc["train", ("wall_time", "loss")]
+        )
+        self.assertAllClose(train_loss_wall_times, np.arange(0, 10))
+        train_accracy_wall_times = list(
+            dataframe.loc["train", ("wall_time", "accuracy")]
+        )
+        self.assertAllClose(train_accracy_wall_times, 2 * np.arange(0, 10))
+        test_loss_wall_times = list(
+            dataframe.loc["test", ("wall_time", "loss")]
+        )
+        self.assertAllClose(test_loss_wall_times, 600 + np.arange(0, 10))
+        test_accracy_wall_times = list(
+            dataframe.loc["test", ("wall_time", "accuracy")]
+        )
+        self.assertAllClose(test_accracy_wall_times, 600 + 2 * np.arange(0, 10))
+
+    def test_get_scalars_with_pivot_table_with_missing_value(self):
+        mock_api_client = mock.Mock()
+
+        def stream_experiment_data(request, **kwargs):
+            self.assertEqual(request.experiment_id, "789")
+            self.assertEqual(kwargs["metadata"], grpc_util.version_metadata())
+            response = export_service_pb2.StreamExperimentDataResponse()
+            response.run_name = "train"
+            response.tag_name = "batch_loss"
+            response.points.steps.append(0)
+            response.points.values.append(0.5)
+            response.points.wall_times.add(seconds=0, nanos=0)
+            response.points.steps.append(1)
+            response.points.values.append(0.25)
+            response.points.wall_times.add(seconds=1, nanos=0)
+            yield response
+            response = export_service_pb2.StreamExperimentDataResponse()
+            response.run_name = "train"
+            response.tag_name = "epoch_loss"
+            response.points.steps.append(0)
+            response.points.values.append(0.375)
+            response.points.wall_times.add(seconds=2, nanos=0)
+            yield response
+
+        mock_api_client.StreamExperimentData = mock.Mock(
+            wraps=stream_experiment_data
+        )
+
+        stderr_messages = []
+
+        def fake_stderr_write(msg):
+            stderr_messages.append(msg)
+
+        with mock.patch.object(
+            experiment_from_dev,
+            "get_api_client",
+            lambda api_endpoint: mock_api_client,
+        ):
+            experiment = experiment_from_dev.ExperimentFromDev("789")
+            with mock.patch.object(sys.stderr, "write", fake_stderr_write):
```

Comment:
Per other comment, would really prefer to not write to `stderr`, but if
you must then please test this with `contextlib.redirect_stderr` rather
than a mock to at least get the right semantics.